### PR TITLE
[mlir][scf] Extend fuse producer to multi-level candidates case

### PR DESCRIPTION
**Pull request by Yun-Fly**
Although producer can be already fused within nest loop structure at present, it didn't work for multi-level  candidates `extractSliceOp` case, E.g.
```
func.func @gemm_fill_fusion(%arg0: tensor<256x512xf32>, %arg1: tensor<512x256xf32>, %arg2: tensor<256x256xf32>) -> tensor<256x256xf32> {
    %c0 = arith.constant 0 : index
    %c64 = arith.constant 64 : index
    %c128 = arith.constant 128 : index
    %cst = arith.constant 0.000000e+00 : f32
    %dest0 = tensor.empty() : tensor<256x256xf32>
    %dest1 = linalg.fill ins(%cst : f32) outs(%dest0 : tensor<256x256xf32>) -> tensor<256x256xf32>
    %1 = scf.forall (%arg3, %arg4) in (2, 2) shared_outs(%arg5 = %dest1) -> tensor<256x256xf32> {
      %iv0 = affine.apply #map(%arg3)
      %iv1 = affine.apply #map(%arg4)
      %extracted_slice_1 = tensor.extract_slice %arg5[%iv0, %iv1] [128, 128] [1, 1] : tensor<256x256xf32> to tensor<128x128xf32>
      %extracted_slice_2 = tensor.extract_slice %arg0[%iv0, 0] [128, 512] [1, 1] : tensor<256x512xf32> to tensor<128x512xf32>
      %extracted_slice_3 = tensor.extract_slice %arg1[0, %iv1] [512, 128] [1, 1] : tensor<512x256xf32> to tensor<512x128xf32>
      %2 = scf.for %arg6 = %c0 to %c128 step %c64 iter_args(%arg7 = %extracted_slice_1) -> (tensor<128x128xf32>) {
        %3 = scf.for %arg8 = %c0 to %c128 step %c64 iter_args(%arg9 = %arg7) -> (tensor<128x128xf32>) {
          %extracted_slice_4 = tensor.extract_slice %arg9[%arg6, %arg8] [64, 64] [1, 1] : tensor<128x128xf32> to tensor<64x64xf32>
          %extracted_slice_5 = tensor.extract_slice %extracted_slice_2[%arg6, 0] [64, 512] [1, 1] : tensor<128x512xf32> to tensor<64x512xf32>
          %extracted_slice_6 = tensor.extract_slice %extracted_slice_3[0, %arg8] [512, 64] [1, 1] : tensor<512x128xf32> to tensor<512x64xf32>
          %4 = linalg.matmul ins(%extracted_slice_5, %extracted_slice_6 : tensor<64x512xf32>, tensor<512x64xf32>) outs(%extracted_slice_4 : tensor<64x64xf32>) -> tensor<64x64xf32>
          %insert_slice = tensor.insert_slice %4 into %arg9[%arg6, %arg8] [64, 64] [1, 1] : tensor<64x64xf32> into tensor<128x128xf32>
          scf.yield %insert_slice : tensor<128x128xf32>
        }
        scf.yield %3 : tensor<128x128xf32>
      }
      scf.forall.in_parallel {
         tensor.parallel_insert_slice %2 into %arg5[%iv0, %iv1] [128, 128] [1, 1] : tensor<128x128xf32> into tensor<256x256xf32>
      }
    }
    return %1 : tensor<256x256xf32>
  }
```

If we want to fuse producer `fill` into `%extracted_slice_4`, current status in master will be broken at `getUntiledProducerFromSliceSource` due to upper candidate slice `%extracted_slice_1` . This PR extends fusing producer to multi-level  `extractSliceOp` cases.

This patch uses multiple application of existing tiling interface as same as another counterpart [PR](https://github.com/llvm/llvm-project/pull/94190).

### CODE REVIEW

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 1]
```diff
@@ -835,7 +835,7 @@ getUntiledProducerFromSliceSource(OpOperand *source,
 /// Implementation of fusing producer of a single slice by computing the
 /// slice of the producer in-place.
 std::optional<scf::SCFFuseProducerOfSliceResult>
-mlir::scf::tileAndFuseProducerOfSlice(
+mlir::scf::tileAndFuseProducerOfSliceImpl(
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:838]: Does this need to be made a publicly exposed method? It makes me squirm when a method that has "impl' in its name is exposed this way. It seems like it is leaking implementation details.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Thanks! Please give me some time to review. Broadly my concern is how do you control which loop nest you want to fuse into when you are doing multiple tiling levels... I can see it is being controlled by API knobs, but those seem fragile. I need to look more in detail though.

**Comment by Yun-Fly** [comment]
> Thanks! Please give me some time to review. Broadly my concern is how do you control which loop nest you want to fuse into when you are doing multiple tiling levels... I can see it is being controlled by API knobs, but those seem fragile. I need to look more in detail though.

Good question! In plan, this PR just focuses on the functionality extension. In another word, user is responsible for deciding which loop nest to fuse by feeding specific candidate into `tileAndFuseProducerOfSlice` which can accept candidates at any tiling level.

BTW, as introduced in this early [RFC](https://discourse.llvm.org/t/rfc-introduce-new-pass-transform-fusion-by-diffusion/79603), we intend to design a higher-level utility or API to control which loop nest to fuse based on semantic check(_validity_) and cost model(_performance_), both of which is expected to be designed as an call-back option with default behavior.

I can give you more example for details:

```
%0 = tensor.pack ins() out() : tensor<128x256xf32> -> tensor<16x8 x 8x32 xf32>
%1 = scf.for()
   %2 = extract_slice
   %3 = scf.for() 
   %4 = extract_slice
```

Looking forward to hear your advice!

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 1 above)
Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:838]: Yes, good suggestion! I will make it `static` in next commit:)

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
THis seems related to the other PR (https://github.com/llvm/llvm-project/pull/94190). Lets come back to this after we have some idea of that. Or if you want to try to land this first, we can switch the order around.

**Comment by Yun-Fly** [comment]
> THis seems related to the other PR (#94190). Lets come back to this after we have some idea of that. Or if you want to try to land this first, we can switch the order around.

I have no specific request about the priory order so far. Its up to you or which one is closer to land status in your mind. Hopefully both of them would be merged ASAP because another coming PR may depend on them.

**Comment by MaheshRavishankar** [comment]
> > THis seems related to the other PR (#94190). Lets come back to this after we have some idea of that. Or if you want to try to land this first, we can switch the order around.
> 
> I have no specific request about the priory order so far. Its up to you or which one is closer to land status in your mind. Hopefully both of them would be merged ASAP because another coming PR may depend on them.

Ok, lets do the producer fusion first, cause that is easier to navigate than the consumer fusion.

(I know you say ASAP, but this is involved code and reviews take a time. Ill do my best to keep it timely, but I am juggling this with other work that is a bit higher priority for my employer)

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
This PR is much easier to review... Ill ask a few more questions to help navigate this.

Trying to put all of this logic upstream is hard. This seems like a very special case where you already have someone that is partially tiling the code and you are trying to reuse some logic from tile and fuse for this case. But there is too much of an assumption being made that the loop structure you create is consistent with what tiling creates. In fact that assumption is already made by the implementation. That is why the current `mlir::tileAndFuseProducerOfSlice` is taking the loop nest generated by `MutableArrayRef<LoopLikeOpInterface> loops` as arguments. 
I'd like to understand if there is a way you can keep the logic of finding loops closer to your context and then consider if we need to change the implementation of how the producer is found by walking the loop nest and extract slice source definition chain. Does that question make sense?

**Comment by Yun-Fly** [comment]
> But there is too much of an assumption being made that the loop structure you create is consistent with what tiling creates. In fact that assumption is already made by the implementation. 

The loop structure created in this PR is exactly what multiple level tiling creates. So it quite looks consistent with what tiling creates but different in multiple level nest loops and candidate slices for better utilization of parallelism and memory hierarchy. I will detail more our context at last.

> That is why the current `mlir::tileAndFuseProducerOfSlice` is taking the loop nest generated by `MutableArrayRef<LoopLikeOpInterface> loops` as arguments. 

Yes, I know it and that is one of reason why this PR looks easier reviewer than another patch of fusing consumer. IIRC, current `mlir::tileAndFuseProducerOfSlice` is only used in `tileConsumerAndFuseProducersUsingSCF`, where consumer is tiled by `tileUsingSCF` into single level tiling with all extract slices located in inner-most loop. So, it is good enough for this use case.

> I'd like to understand if there is a way you can keep the logic of finding loops closer to your context and then consider if we need to change the implementation of how the producer is found by walking the loop nest and extract slice source definition chain. Does that question make sense?

Our context is that some developers want to costume the way how to tile the certain op instead of directly using general but common `tileUsingSCF` API, especially for `matmul` or `convolution` on different target device. The only concern of these developers is how to improve the kernel performance by multiple level tiling, but not how subsequent fusion would be done because what kind of post ops are uncertain for them. In another word, this kind of use case decouples the tiling stage of contraction op and fusion process of surrounding ops, which brings more flexible tiling and requires robust fusion interface, saying `tileAndFuseProducerOfSlice`.

**Comment by Yun-Fly** [comment]
> I know you say ASAP, but this is involved code and reviews take a time. Ill do my best to keep it timely.

I would like to express my sincere gratitude for your involving!

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 2]
```diff
@@ -949,6 +949,145 @@ mlir::scf::tileAndFuseProducerOfSlice(
...
+///   %2 = extract %arg1
+///   %3 = scf.for(%arg2 = %2)
+///      %4 = extract %args2
+///      ...
+/// ```
+///
+/// @param candidateSliceOp: %4 = extract %args2
+/// @param backwardSlice: in-out parameter populated by backward extractSliceOps
+/// @return OpResult Producer : %0 = producer
+static FailureOr<OpResult> getRealProducerFromExtractSliceOp(
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:966]: Instead of doing this, would it help if you just folded the `extract_slices` using something like [this](https://github.com/iree-org/llvm-project/blob/c9f6518f742c88bda309d5331e0a5d4664387f94/mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp#L23) pattern

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 3]
```diff
@@ -949,6 +949,145 @@ mlir::scf::tileAndFuseProducerOfSlice(
...
+///  %0 = scf.for()
+///    %1 = scf.for()
+///      %2 = scf.for()
+/// ```
+///
+/// If `%2 = scf.for` is given without specific prediction function, this
+/// function will return three nest loops: %0 + %1 + %2.
+static SmallVector<LoopLikeOpInterface> getOuterNestLoopsWhile(
+    LoopLikeOpInterface loop,
+    const std::function<LogicalResult(LoopLikeOpInterface)> &pred) {
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1020]: The indirection from the callback is confusing. Maybe just "inline" the call back into this method.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp [hunk 4]
```diff
@@ -949,6 +949,145 @@ mlir::scf::tileAndFuseProducerOfSlice(
...
+    return std::nullopt;
+  }
+
+  std::optional<scf::SCFFuseProducerOfSliceResult> fuseProducerResult;
+  // reverse from outer to inner
+  std::reverse(backwardSlice.begin(), backwardSlice.end());
+  // multiple application of `tileAndFuseProducerOfSliceImpl`
+  for (auto &&[index, sliceOp] : llvm::enumerate(backwardSlice)) {
+    // get nest loops between next candidate sliceOp and tiled producer.
+    auto whileProducerOutOfLoopBlock =
```
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1060]: This is not enough. You need to check that the `extract_slice` is in the innermost loop of the oop nest generated by tiling. The reason is that when you are yielded the value of a fused producer that yielding only works (AFAIK)
1) The innermost loop returns the value
2) All surrounding outer loops are just yielding the result of the inner loops (and that the loop nest just pass the init values along from the outer loop to the inner loop.

So if we want that to still hold, then this loop nest needs to be checked that way...

**Review by MaheshRavishankar — COMMENTED** [review]
Left a few more comments. Thanks for being patient. I am a bit busy this week, but I am more free next week. If you are on discord (IREE discord or MLIR discord), we can connect there and I am happy to work with you with a faster cadence to help you land what you need upstream, and maybe suggest things you can keep downstream since it is valid only for your context.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 2 above)
Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:966]: I know that we can merge consecutive `extractSlice`. But, I am sorry that we have to just keep these consecutive `extractSlice` for selection. The most intuitive example is as below:

```
%0 = producer
%1 = scf.for(%arg1 = %0)
   %2 = extract %arg1
   %3 = scf.for(%arg2 = %2)
      %4 = extract %args2
```
In some cases, the producer could be only fused into the outer candidate slice due to semantic of producer, i.e. reduce/pack/unpack those ops have specific demands of tiling size. If we merge the two consecutive `extractSlice`, it may break this fusion. And that is why we emphasize multi-level candidates here.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 3 above)
Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1020]: Like `ArrayRef()::take_while(PredicateT Pred)` which returns the first N elements of this Array that satisfy the given predicate, this callback will help to filter out which kinds of outer loops we focus on.

Besides, it is a shared utility also used in another fusing producer patch([here](https://github.com/llvm/llvm-project/pull/94190/files#diff-088c31bdb684f78d48f7607d0007daa4e1fba53b0361def9841f79631d8a3d5dR1278)), so it maybe better to keep this callback for reuse.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 4 above)
Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1060]: I am afraid that I can't get it. Could you illustrate this with an concrete example? Thanks!

**Comment by Yun-Fly** [comment]
> Thanks for being patient. I am a bit busy this week, but I am more free next week. If you are on discord (IREE discord or MLIR discord), we can connect there and I am happy to work with you with a faster cadence...

Yes, surely. I just ping you there. Feel free to have a talk if you have time.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 3 above)
Comment (Menooker) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1020]: BTW, we might be able to use function_ref here, as long as we don't need to remember the function `pred` outside of the function.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:966]: That is OK, but the point here is to fuse with `%2` you dont need to look through `extract_slices`.... Just not sure of why there would be a case where you need to look through extract_slices to get to the producer, when you cant collapse all the extract slices to begin with.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 4 above)
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1060]: Lets say you are starting with this
```
%0 = scf.for ... (%arg0 = %1) {
    %2 = scf.for ... (%arg1 = %arg0) {
        %3 ... scf.for ... (%arg2 = %arg1 ) {
          %4 ....
          %5...
           scf.yield %4
        }
        scf.yield %3
    }
    scf.yield %2
}    
```

Now if you want to yield `%5` on top of that the logic only works if `%5` is in the inner-most loop and the loop defining `%2` is yielding the result of loop defining `%3`, and loop defining `%0` is yielding the result of the loop defining `%2` (and so on...)

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 2 above)
Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:966]: That is just simple test. If we have one more candidate slice, saying:

```
%0 = producer
%1 = scf.for(%arg1 = %0)
   %2 = extract %arg1
   %3 = scf.for(%arg2 = %2)
      %4 = extract %args2
      %5 = scf.for(%arg3 = %4)
          %6 = extract %args3
          tiled_consumer ins(%6)
```
where `%2` and `%4` are both valid candidate to fuse producer. And the user just want to manually pass `%4` to `tileAndFuseProducerOfSlice` because `%4` usually has smaller tileSize in most cases. Then, the current `getUntiledProducerFromSliceSource` will fail to get real producer without looking through `extract_slices`.

Another use case is that if someone want to enable automatic fusion by starting with `tiled_consumer` just like what `tileConsumerAndFuseProducersUsingSCF` did. The available extract slice we can find via operand of `tiled_consumer` is only the nearest candidate `%6`. Then, we also need to look through all `extract_slices` to get real producer.

Meanwhile, we dont know which candidate slice could be fuse until we know what the exact producer is. Thus, it is unreasonable to collapse all the extract slices before fusion.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 4 above)
Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1311]: Ok, I see. Thanks for this reminder. I have add additional check called `isForOpYieldResultOfInnerLoop`. BTW, it is also shared in another fusing consumer [PR](https://github.com/llvm/llvm-project/pull/94190).

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1202]: I am not sure this reasoning holds. If you want to fuse `%4` and not `%2` you are effectively "fusing" the slices and the doing the producer fusion. It would be better to stage it so that you fuse the slices so that you get the producer directly and then do the fusion. So effectively this change is trying to do both of this
1. Combine extract slices
2. Do producer fusion
in code. That is just added complexity. It would be easier to just combine the extract slices first and then fuse the producer.

**Review by MaheshRavishankar — COMMENTED** [review]
I dont know if this commit is still relevant. I was looking for a commit where things were broken up a bit more. If this commit is still relevant, I responded to the outstanding comment here.

File: mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp (hunk 2 above)
Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1202]: Let me try to rephrase the whole process step by step:
1. look through to collect all backward candidate slice of tiled operand `%6`, including `%6 = extract %args3`, ` %4 = extract %args2`, `%2 = extract %arg1`.
2. select and filter which candidate to fuse. I.e. ` %4 = extract %args2`.
3. Iteratively do producer fusion, from `%2` to `%4` by two iterations.

Through second thought, I agree that we can replace `step3` with your advice, which firstly combined `%2` and `%4`, then execute only single iteration of producer fusion.

However, as for `step 1` and `step 2`, we need to look through extract slice anyway... As I explained early, we could not fold `%2` and `%4` before these two steps. 

BTW, I dont think original `step3` in iterative fashion adding too much complexity because it is just reusing same logic and code.

On the other side, merging two extract slice in our case seems a little different from [MergeConsecutiveExtractSlice](https://github.com/iree-org/llvm-project/blob/c9f6518f742c88bda309d5331e0a5d4664387f94/mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp#L23) as you pointed out, which actually expects direct use-def relationship:

```
%0 = extract_slice
%1 = extract_slice %0
```
rather than indirect relationship in our example:

```
%0 = extract_slice
scf.for(%1=%0)
  %2 = extract_slice %1
```

In this case, it is also necessary to look through extract slice before we merge these two slices.

Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1202]:
```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%2)
      %4 = extract_slice %3
```

Besides, I am worried about merging two candidates interleaved by a `scf.for` with its `inits` involved is not trivial...

Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1202]: Hi, @MaheshRavishankar. Since PR #108318 has been broken down into three parts:
1. support single nested `scf.for` (merged, great thanks for your review!)
2. multi-level candidates
3. multiple users

I think the second part is similar to this patch, no matter producer or consumer fusion. In general, this patch focus on how to deal with consecutive candidates **interleaved by a `scf.for`**.

Lets look back what the major argument remains so far to speed up our next stage of review:
1. why need to look through the total chain of candidate slice?
2. why not merge two consecutive candidates?

The answer is listed in above several threads. The root cause behind are two different solutions to solve multi-level candidates cases:
1. merge consecutive candidates in advance.
2. iteratively fuse producer/consumer with one of candidates step by step(from outer to inner).

IMO, the most concern about the first solution is that it may introduce too many additional transform regarding `scf.for`. E.g..

```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%2)
      %4 = extract_slice %3
      ... 
      yield %x
```

As I explained above, this scenario is a little different from what [`MergeConsecutiveExtractSlice`](https://github.com/iree-org/llvm-project/blob/c9f6518f742c88bda309d5331e0a5d4664387f94/mlir/lib/Dialect/Tensor/Transforms/MergeConsecutiveInsertExtractSlicePatterns.cpp#L23) expects due to `scf.for`. 

If we just merge consecutive extract slice(BTW, here we also need look through to penetrate `scf.for` ), the possible resultant IR looks like below:
```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%2)
      %4 = extract_slice %1 [newOffsets] [newSizes]
      ... 
      yield %x
```
Then the problem is that how we modify new `inits` of `scf.for` with latest `dpsInit` of fused `producerOp`? ([`getUntiledProducerFromSliceSource`](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp#L1058) will fail to find `destinationIterArg`)

On the other hand, if we modify `%3=%2` at the same time, the resultant IR would become:
```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%1)
      %4 = extract_slice %3 [newOffsets] [newSizes]
      ... 
      yield %x_new
```
First of all, this transform has already introduced more code change and there is an assumption that `%3` has only single one user. Moreover, considering the semantic of `scf.for`, it seems that we also need to modify yield value `%x`(and even its producer, like `insert_slice`) to make the result of `scf.for` matched with its `inits` . 

In summary, I am afraid that this kind of complexity is much more than the second solution based on simple iterative process of reusing existed function.

Comment (MaheshRavishankar) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1202]: Thanka for the explanation. I see the issues a bit more clearly now. I havent reviewed the code yet. I dont know if this has been rebased on top of what was submitted. 
But here is the complexity that I am concerned about. In you example, you need to somewhere keep track of the sequence of extract slices that you need to walk to get the producer because the actual offset and size you need is obtained by "combining" the offsets and sizes of all the slices to get the "real offset" and size.  Also I am not convinced that fusing the first extract slice + producer and then doing the second extract slice + producer is not feasible. That should be always possible.

So if you start with this

```
%0 = producerOp
scf.for(%1=%0)
   %2 = extract_slice %1
   scf.for(%3=%2)
      %4 = extract_slice %3
      ... 
      yield %x
```

you should always be able to do

```
scf.for(%1=%0)
   %2 = tiled producer 1
   scf.for(%3=%2)
      %4 = extract_slice %3
      ... 
      yield %x
```

and then you do 

```
scf.for(%1=%0)
   scf.for(%3=%2)
      %4 = tiled producer 2
      ... 
      yield %x
```

The amount of state you need to carry during the transformation to fuse with one extract slice is prohibitively high IMO. That will make it hard to change/fix the transformation.

Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1202]: > I dont know if this has been rebased on top of what was submitted.

Not rebased yet, because I think it is more import to reach agreement on solution before pushing...

> you need to somewhere keep track of the sequence of extract slices that you need to walk to get the producer because the actual offset and size you need is obtained by "combining" the offsets and sizes of all the slices to get the "real offset" and size.

Yes, from the view of consequentialism, the real offset is indeed combined. However, there are something else we need to address, like what I explained above regarding loop transform. In this way, each two of extract_slice  (interleaved by a `scf.for`) merged, the certain `scf.for`(and its yieldValue, etc...) need transformed. Is the amount of state need to carry similar to what I do in iterative fashion?

> Also I am not convinced that fusing the first extract slice + producer and then doing the second extract slice + producer is not feasible. That should be always possible.

It is decided by concrete semantic of tilable op, such as `reduce`/`pack`/`unpack` op. Lets say fusing producer `unpack` into following loop:

```
// unpack tensor from ABab to AB
%1 = tensor.unpack ... inner_tiles = [32, 32] ... : tensor<2x2x32x32xbf16> -> tensor<64x64xbf16>
scf.for(0, 2) { // tile A dimension
 extract_slice1 // tileSize<32, 64>
 scf.for(0, 4) { // tile B dimension
    extract_slice2 // tileSize<32, 16>
    ...
 }
}
```
As you can see, the `tileSize` comes from `extract_slice2` is `<32,16>`, but `tensor.unpack` prefer perfect tiling case, i.e. tileSize should be exactly divided by `inner_tiles`. So, it may be not feasible in this case.

BTW, fusing consumer `reduce` maybe more intuitive: 

```
%0 = scf.for(0, 2) { // tile 128 dimension
 scf.for(0, 4) { // tile B dimension
    ...
   insert_slice1 // tileSize<64, 64>
 }
 insert_slice1 // tileSize<64, 256>
}
%2 = linalg.reduce { arith.addf } ins(%0 : tensor<128x256xf32>) outs(%1: tensor<128xf32>) dimensions = [1]
```
We could not furtherly fuse `reduce` into the inner-most insert slice, otherwise, it will lead to partial reduce(its another topic).


> So if you start with this

> you should always be able to do

> and then you do

Yes, it is exactly what I do now. IIRC, that is also what you suggest before in nested consumer fusion...

Comment (Yun-Fly) [reply mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp:1202]: Ping
